AudioDeviceExample/AudioDevices/ExampleAVAudioEngineDevice.m

```diff
@@ -0,0 +1,810 @@
+            if (numFrames > context->maxFramesPerBuffer) {
+                NSLog(@"Can handle a max of %u frames but got %u.", (unsigned int)context->maxFramesPerBuffer, (unsigned int)numFrames);
+            }
+            *actionFlags |= kAudioUnitRenderAction_OutputIsSilence;
+            memset(audioBuffer, 0, audioBufferSizeInBytes);
+        }
+    }
+
+    return noErr;
+}
+
+static OSStatus ExampleAVAudioEngineDeviceRecordCallback(void *refCon,
+                                                         AudioUnitRenderActionFlags *actionFlags,
+                                                         const AudioTimeStamp *timestamp,
+                                                         UInt32 busNumber,
+                                                         UInt32 numFrames,
+                                                         AudioBufferList *bufferList) {
+
+    if (numFrames > kMaximumFramesPerBuffer) {
+        NSLog(@"Expected %u frames but got %u.", (unsigned int)kMaximumFramesPerBuffer, (unsigned int)numFrames);
+        return noErr;
+    }
+
+    AudioCapturerContext *context = (AudioCapturerContext *)refCon;
+
+    if (context->deviceContext == NULL) {
+        return noErr;
+    }
+
+    AudioBufferList *audioBufferList = context->bufferList;
+    audioBufferList->mBuffers[0].mDataByteSize = numFrames * sizeof(UInt16) * kPreferredNumberOfChannels;
+    audioBufferList->mBuffers[0].mData = NULL;
+
+    OSStatus status = noErr;
+    status = AudioUnitRender(context->audioUnit,
+                             actionFlags,
+                             timestamp,
+                             1,
+                             numFrames,
+                             audioBufferList);
+
+    int8_t *audioBuffer = (int8_t *)audioBufferList->mBuffers[0].mData;
+    UInt32 audioBufferSizeInBytes = audioBufferList->mBuffers[0].mDataByteSize;
+
+    if (context->deviceContext) {
+        TVIAudioDeviceWriteCaptureData(context->deviceContext, audioBuffer, audioBufferSizeInBytes);
+    }
+
+    return noErr;
+}
+
+#pragma mark - Private (AVAudioSession and CoreAudio)
+
++ (nullable TVIAudioFormat *)activeFormat {
+    /*
+     * Use the pre-determined maximum frame size. AudioUnit callbacks are variable, and in most sitations will be close
+     * to the `AVAudioSession.preferredIOBufferDuration` that we've requested.
+     */
+    const size_t sessionFramesPerBuffer = kMaximumFramesPerBuffer;
+    const double sessionSampleRate = [AVAudioSession sharedInstance].sampleRate;
+    const NSInteger sessionOutputChannels = [AVAudioSession sharedInstance].outputNumberOfChannels;
+    size_t rendererChannels = sessionOutputChannels >= TVIAudioChannelsStereo ? TVIAudioChannelsStereo : TVIAudioChannelsMono;
+
+    return [[TVIAudioFormat alloc] initWithChannels:rendererChannels
+                                         sampleRate:sessionSampleRate
+                                    framesPerBuffer:sessionFramesPerBuffer];
+}
```

Comment:
Should there be two factory methods, one for capturing and one for rendering formats? If there is a mismatch, the rendering format determines the recording format, and we will let the audio graph to do the conversion for us.

Reply:
We decide not to because the entire session has one sample rate and we always ask for mono.